ambry-protocol/src/main/java/com.github.ambry.protocol/MessageInfoAndMetdataListSerde.java

```diff
@@ -0,0 +1,223 @@
+          size += 1;
+          if (messageInfo.getCrc() != null) {
+            // crc
+            size += Long.BYTES;
+          }
+          break;
+        case VERSION_3:
+          // whether crc is present
+          size += 1;
+          if (messageInfo.getCrc() != null) {
+            // crc
+            size += Long.BYTES;
+          }
+          // accountId
+          size += Short.BYTES;
+          // containerId
+          size += Short.BYTES;
+          // operationTime
+          size += Long.BYTES;
+          break;
+        case VERSION_4:
+          // whether crc is present
+          size += 1;
+          if (messageInfo.getCrc() != null) {
+            // crc
+            size += Long.BYTES;
+          }
+          // accountId
+          size += Short.BYTES;
+          // containerId
+          size += Short.BYTES;
+          // operationTime
+          size += Long.BYTES;
+          if (messageMetadata != null) {
+            size += Integer.BYTES;
+            size += messageMetadata.sizeInBytes();
+          } else {
+            size += Integer.BYTES;
+          }
+          break;
+
+        default:
+          throw new IllegalArgumentException("Unknown version in MessageInfoList " + version);
+      }
+    }
+    return size;
+  }
+
+  void serializeMessageInfoAndMetadataList(ByteBuffer outputBuffer) {
+    outputBuffer.putInt(messageInfoList == null ? 0 : messageInfoList.size());
+    if (messageInfoList != null) {
+      ListIterator<MessageInfo> infoListIterator = messageInfoList.listIterator();
+      ListIterator<MessageMetadata> metadataListIterator = messageMetadataList.listIterator();
+      while (infoListIterator.hasNext()) {
+        MessageInfo messageInfo = infoListIterator.next();
+        MessageMetadata messageMetadata = metadataListIterator.next();
+        outputBuffer.put(messageInfo.getStoreKey().toBytes());
+        outputBuffer.putLong(messageInfo.getSize());
+        outputBuffer.putLong(messageInfo.getExpirationTimeInMs());
+        outputBuffer.put(messageInfo.isDeleted() ? DELETED : (byte) ~DELETED);
+        if (version < VERSION_1 || version > VERSION_4) {
+          throw new IllegalArgumentException("Unknown version in MessageInfoList " + version);
+        }
+        if (version > VERSION_1) {
+          Long crc = messageInfo.getCrc();
+          if (crc != null) {
+            outputBuffer.put(CRC_PRESENT);
+            outputBuffer.putLong(crc);
+          } else {
+            outputBuffer.put((byte) ~CRC_PRESENT);
+          }
+        }
+        if (version > VERSION_2) {
+          outputBuffer.putShort(messageInfo.getAccountId());
+          outputBuffer.putShort(messageInfo.getContainerId());
+          outputBuffer.putLong(messageInfo.getOperationTimeMs());
+        }
+        if (version > VERSION_3) {
+          if (messageMetadata == null) {
+            outputBuffer.putInt(0);
+          } else {
+            messageMetadata.serializeMessageMetadata(outputBuffer);
```

Comment:
shouldn't you add size of MessageMetadata to the outputBuffer before serializing the MessageMetadata

Reply:
Yeah, this is incorrect, thanks for catching. Seems like this class never had any tests (this is not a newly introduced class). I have added a test to catch this scenario.